Honor .npmrc proxy settings

### OS:
OSX Sierra

### Versions:

```
angular-cli: 1.0.0-beta.24
node: 4.4.0
```

### Reproduce steps:
I used the [angular cli](https://github.com/angular/angular-cli) to generate a project and run the default e2e test that comes out of the box. __Keep in mind, I did this while on a network that blocks access to ...googleapis.com__ 
```
ng new PROJECT_NAME
cd PROJECT_NAME
ng serve
ng e2e
```

### The log given by the failure:
```

> temp-project-name@0.0.0 pree2e /Users/user-name/Projects/temp-project-name
> webdriver-manager update --standalone false --gecko false


events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: getaddrinfo ENOTFOUND chromedriver.storage.googleapis.com chromedriver.storage.googleapis.com:443
    at errnoException (dns.js:26:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:77:26)

npm ERR! Darwin 16.0.0
npm ERR! argv "/usr/local/Cellar/node4-lts/4.4.0/bin/node" "/Users/user-name/.npm-packages/bin/npm" "run" "e2e" "--" "./protractor.conf.js"
npm ERR! node v4.4.0
npm ERR! npm  v3.8.0
npm ERR! code ELIFECYCLE
npm ERR! temp-project-name@0.0.0 pree2e: `webdriver-manager update --standalone false --gecko false`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the temp-project-name@0.0.0 pree2e script 'webdriver-manager update --standalone false --gecko false'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the temp-project-name package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     webdriver-manager update --standalone false --gecko false
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs temp-project-name
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls temp-project-name
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/user-name/Projects/temp-project-name/npm-debug.log
```

### Current workaround
I set the http_proxy and https_proxy environment variables to address this issue, but it would be a much better experience if .npmrc global/local proxy settings were honored instead.

--- a/lib/provider/appium.ts
+++ b/lib/provider/appium.ts
@@ -5,31 +5,24 @@ import * as path from 'path';
 import * as rimraf from 'rimraf';
 import * as semver from 'semver';
 
-import {ProviderConfig, ProviderInterface} from './provider';
+import {ProviderClass, ProviderConfig, ProviderInterface} from './provider';
 import {requestBody} from './utils/http_utils';
 
 const log = loglevel.getLogger('webdriver-manager');
 
-export class Appium implements ProviderInterface {
+export class Appium extends ProviderClass implements ProviderInterface {
   ignoreSSL: boolean;
   outDir: string;
   outDirAppium: string;
   proxy: string;
   requestUrl = 'http://registry.npmjs.org/appium';
 
-  constructor(providerConfig?: ProviderConfig) {
-    if (providerConfig) {
-      this.ignoreSSL = providerConfig.ignoreSSL;
-      if (providerConfig.outDir) {
-        this.outDir = providerConfig.outDir;
-      }
-      if (providerConfig.proxy) {
-        this.proxy = providerConfig.proxy;
-      }
-      if (providerConfig.requestUrl) {
-        this.requestUrl = providerConfig.requestUrl;
-      }
-    }
+  constructor(config?: ProviderConfig) {
+    super();
+    this.ignoreSSL = this.setVar('ignoreSSL', this.ignoreSSL, config);
+    this.outDir = this.setVar('outDir', this.outDir, config);
+    this.proxy = this.setVar('proxy', this.proxy, config);
+    this.requestUrl = this.setVar('requestUrl', this.requestUrl, config);
   }
 
   /**

--- a/lib/provider/chromedriver.ts
+++ b/lib/provider/chromedriver.ts
@@ -2,13 +2,13 @@ import * as fs from 'fs';
 import * as os from 'os';
 import * as path from 'path';
 
-import {OUT_DIR, ProviderConfig, ProviderInterface} from './provider';
+import {OUT_DIR, ProviderClass, ProviderConfig, ProviderInterface} from './provider';
 import {convertXmlToVersionList, updateXml} from './utils/cloud_storage_xml';
 import {changeFilePermissions, generateConfigFile, getBinaryPathFromConfig, removeFiles, renameFileWithVersion, unzipFile, zipFileList,} from './utils/file_utils';
 import {requestBinary} from './utils/http_utils';
 import {getVersion} from './utils/version_list';
 
-export class ChromeDriver implements ProviderInterface {
+export class ChromeDriver extends ProviderClass implements ProviderInterface {
   cacheFileName = 'chromedriver.xml';
   configFileName = 'chromedriver.config.json';
   ignoreSSL = false;
@@ -19,31 +19,16 @@ export class ChromeDriver implements ProviderInterface {
   requestUrl = 'https://chromedriver.storage.googleapis.com/';
   seleniumFlag = '-Dwebdriver.chrome.driver';
 
-  constructor(providerConfig?: ProviderConfig) {
-    if (providerConfig) {
-      if (providerConfig.cacheFileName) {
-        this.cacheFileName = providerConfig.cacheFileName;
-      }
-      if (providerConfig.configFileName) {
-        this.configFileName = providerConfig.configFileName;
-      }
-      this.ignoreSSL = providerConfig.ignoreSSL;
-      if (providerConfig.osArch) {
-        this.osArch = providerConfig.osArch;
-      }
-      if (providerConfig.osType) {
-        this.osType = providerConfig.osType;
-      }
-      if (providerConfig.outDir) {
-        this.outDir = providerConfig.outDir;
-      }
-      if (providerConfig.proxy) {
-        this.proxy = providerConfig.proxy;
-      }
-      if (providerConfig.requestUrl) {
-        this.requestUrl = providerConfig.requestUrl;
-      }
-    }
+  constructor(config?: ProviderConfig) {
+    super();
+    this.cacheFileName = this.setVar('cacheFileName', this.cacheFileName, config);
+    this.configFileName = this.setVar('configFileName', this.configFileName, config);
+    this.ignoreSSL = this.setVar('ignoreSSL', this.ignoreSSL, config);
+    this.osArch = this.setVar('osArch', this.osArch, config);
+    this.osType = this.setVar('osType', this.osType, config);
+    this.outDir = this.setVar('outDir', this.outDir, config);
+    this.proxy = this.setVar('proxy', this.proxy, config);
+    this.requestUrl = this.setVar('requestUrl', this.requestUrl, config);
   }
 
   /**

--- a/lib/provider/geckodriver.ts
+++ b/lib/provider/geckodriver.ts
@@ -2,7 +2,7 @@ import * as fs from 'fs';
 import * as os from 'os';
 import * as path from 'path';
 
-import {OUT_DIR, ProviderConfig, ProviderInterface} from './provider';
+import {OUT_DIR, ProviderClass, ProviderConfig, ProviderInterface} from './provider';
 import {changeFilePermissions, generateConfigFile, getBinaryPathFromConfig, removeFiles, renameFileWithVersion, tarFileList, uncompressTarball, unzipFile, zipFileList,} from './utils/file_utils';
 import {convertJsonToVersionList, updateJson} from './utils/github_json';
 import {requestBinary} from './utils/http_utils';
@@ -12,7 +12,7 @@ export interface GeckoDriverProviderConfig extends ProviderConfig {
   oauthToken?: string;
 }
 
-export class GeckoDriver implements ProviderInterface {
+export class GeckoDriver extends ProviderClass implements ProviderInterface {
   cacheFileName = 'geckodriver.json';
   configFileName = 'geckodriver.config.json';
   ignoreSSL = false;
@@ -24,34 +24,17 @@ export class GeckoDriver implements ProviderInterface {
   requestUrl = 'https://api.github.com/repos/mozilla/geckodriver/releases';
   seleniumFlag = '-Dwebdriver.gecko.driver';
 
-  constructor(providerConfig?: GeckoDriverProviderConfig) {
-    if (providerConfig) {
-      if (providerConfig.cacheFileName) {
-        this.cacheFileName = providerConfig.cacheFileName;
-      }
-      if (providerConfig.configFileName) {
-        this.configFileName = providerConfig.configFileName;
-      }
-      this.ignoreSSL = providerConfig.ignoreSSL;
-      if (providerConfig.osArch) {
-        this.osArch = providerConfig.osArch;
-      }
-      if (providerConfig.osType) {
-        this.osType = providerConfig.osType;
-      }
-      if (providerConfig.outDir) {
-        this.outDir = providerConfig.outDir;
-      }
-      if (providerConfig.proxy) {
-        this.proxy = providerConfig.proxy;
-      }
-      if (providerConfig.requestUrl) {
-        this.requestUrl = providerConfig.requestUrl;
-      }
-      if (providerConfig.oauthToken) {
-        this.oauthToken = providerConfig.oauthToken;
-      }
-    }
+  constructor(config?: GeckoDriverProviderConfig) {
+    super();
+    this.cacheFileName = this.setVar('cacheFileName', this.cacheFileName, config);
+    this.configFileName = this.setVar('configFileName', this.configFileName, config);
+    this.ignoreSSL = this.setVar('ignoreSSL', this.ignoreSSL, config);
+    this.osArch = this.setVar('osArch', this.osArch, config);
+    this.osType = this.setVar('osType', this.osType, config);
+    this.outDir = this.setVar('outDir', this.outDir, config);
+    this.proxy = this.setVar('proxy', this.proxy, config);
+    this.requestUrl = this.setVar('requestUrl', this.requestUrl, config);
+    this.oauthToken = this.setVar('oauthToken', this.oauthToken, config);
   }
 
   /**

--- a/lib/provider/iedriver.ts
+++ b/lib/provider/iedriver.ts
@@ -2,13 +2,13 @@ import * as fs from 'fs';
 import * as os from 'os';
 import * as path from 'path';
 
-import {OUT_DIR, ProviderConfig, ProviderInterface,} from './provider';
+import {OUT_DIR, ProviderClass, ProviderConfig, ProviderInterface} from './provider';
 import {convertXmlToVersionList, updateXml,} from './utils/cloud_storage_xml';
 import {generateConfigFile, getBinaryPathFromConfig, removeFiles, renameFileWithVersion, unzipFile, zipFileList,} from './utils/file_utils';
 import {requestBinary} from './utils/http_utils';
 import {getVersion} from './utils/version_list';
 
-export class IEDriver implements ProviderInterface {
+export class IEDriver extends ProviderClass implements ProviderInterface {
   cacheFileName = 'iedriver.xml';
   configFileName = 'iedriver.config.json';
   ignoreSSL = false;
@@ -19,31 +19,16 @@ export class IEDriver implements ProviderInterface {
   requestUrl = 'https://selenium-release.storage.googleapis.com/';
   seleniumFlag = '-Dwebdriver.ie.driver';
 
-  constructor(providerConfig?: ProviderConfig) {
-    if (providerConfig) {
-      if (providerConfig.cacheFileName) {
-        this.cacheFileName = providerConfig.cacheFileName;
-      }
-      if (providerConfig.configFileName) {
-        this.configFileName = providerConfig.configFileName;
-      }
-      this.ignoreSSL = providerConfig.ignoreSSL;
-      if (providerConfig.osArch) {
-        this.osArch = providerConfig.osArch;
-      }
-      if (providerConfig.osType) {
-        this.osType = providerConfig.osType;
-      }
-      if (providerConfig.outDir) {
-        this.outDir = providerConfig.outDir;
-      }
-      if (providerConfig.proxy) {
-        this.proxy = providerConfig.proxy;
-      }
-      if (providerConfig.requestUrl) {
-        this.requestUrl = providerConfig.requestUrl;
-      }
-    }
+  constructor(config?: ProviderConfig) {
+    super();
+    this.cacheFileName = this.setVar('cacheFileName', this.cacheFileName, config);
+    this.configFileName = this.setVar('configFileName', this.configFileName, config);
+    this.ignoreSSL = this.setVar('ignoreSSL', this.ignoreSSL, config);
+    this.osArch = this.setVar('osArch', this.osArch, config);
+    this.osType = this.setVar('osType', this.osType, config);
+    this.outDir = this.setVar('outDir', this.outDir, config);
+    this.proxy = this.setVar('proxy', this.proxy, config);
+    this.requestUrl = this.setVar('requestUrl', this.requestUrl, config);
   }
 
   /**

--- a/lib/provider/provider.ts
+++ b/lib/provider/provider.ts
@@ -5,7 +5,7 @@ import * as path from 'path';
 export const OUT_DIR = path.resolve(__dirname, '..', '..', '..', 'downloads');
 
 /**
- * The provider updateBinary interface implemented by all providers.
+ * The provider interface implemented by all providers.
  */
 export interface ProviderInterface {
   cleanFiles?: () => string;
@@ -14,6 +14,34 @@ export interface ProviderInterface {
   updateBinary: (version?: string) => Promise<void>;
   seleniumFlag?: string;
   osType?: string;
+  versionConfig?: string;
+}
+
+/**
+ * The provider class with commonmethods by all providers.
+ */
+export class ProviderClass {
+  /**
+   * Setting values from either the npmrc or config object:
+   *   1. Use the default value
+   *   2. If the npmrc value exists, use the npmrc value
+   *   2. If the value is set in the config, use that instead.
+   * @param key 
+   * @param defaultValue
+   * @param providerConfig
+   * @return The value of the variable. The type needs to be fixed.
+   */
+  setVar<T>(key: string, defaultValue: T,
+    providerConfig: ProviderConfig): T {
+  let value: T = defaultValue;
+  if (process.env[`npm_config_${key}`]) {
+    value = process.env[`npm_config_${key}`] as any as T;
+  }
+  if (providerConfig && providerConfig[key]) {
+    value = providerConfig[key] as any as T;
+  }
+  return value;
+  }
 }
 
 /**
@@ -41,6 +69,8 @@ export interface ProviderConfig {
   proxy?: string;
   // Set the requests to ignore SSL (optional).
   ignoreSSL?: boolean;
+  // Allow to pass a config for the max versions to download (optional).
+  versionConfig?: string;
   // Catch all for other things.
   [key: string]: string|boolean|number;
 }

--- a/lib/provider/selenium_server.ts
+++ b/lib/provider/selenium_server.ts
@@ -5,7 +5,7 @@ import * as os from 'os';
 import * as path from 'path';
 import * as request from 'request';
 
-import {OUT_DIR, ProviderConfig, ProviderInterface,} from './provider';
+import {OUT_DIR, ProviderClass, ProviderConfig, ProviderInterface} from './provider';
 import {convertXmlToVersionList, updateXml} from './utils/cloud_storage_xml';
 import {generateConfigFile, getBinaryPathFromConfig, removeFiles,} from './utils/file_utils';
 import {curlCommand, initOptions, requestBinary} from './utils/http_utils';
@@ -19,7 +19,7 @@ export interface SeleniumServerProviderConfig extends ProviderConfig {
   runAsDetach?: boolean;
 }
 
-export class SeleniumServer implements ProviderInterface {
+export class SeleniumServer extends ProviderClass implements ProviderInterface {
   cacheFileName = 'selenium-server.xml';
   configFileName = 'selenium-server.config.json';
   ignoreSSL = false;
@@ -33,40 +33,21 @@ export class SeleniumServer implements ProviderInterface {
   runAsNode = false;
   runAsDetach = false;
 
-  constructor(providerConfig?: SeleniumServerProviderConfig) {
-    if (providerConfig) {
-      if (providerConfig.cacheFileName) {
-        this.cacheFileName = providerConfig.cacheFileName;
-      }
-      if (providerConfig.configFileName) {
-        this.configFileName = providerConfig.configFileName;
-      }
-      this.ignoreSSL = providerConfig.ignoreSSL;
-      if (providerConfig.osArch) {
-        this.osArch = providerConfig.osArch;
-      }
-      if (providerConfig.osType) {
-        this.osType = providerConfig.osType;
-      }
-      if (providerConfig.outDir) {
-        this.outDir = providerConfig.outDir;
-      }
-      if (providerConfig.port) {
-        this.port = providerConfig.port;
-      }
-      if (providerConfig.proxy) {
-        this.proxy = providerConfig.proxy;
-      }
-      if (providerConfig.requestUrl) {
-        this.requestUrl = providerConfig.requestUrl;
-      }
-      if (providerConfig.runAsNode) {
-        this.runAsNode = providerConfig.runAsNode;
-      }
-      if (providerConfig.runAsDetach) {
-        this.runAsDetach = providerConfig.runAsDetach;
-        this.runAsNode = true;
-      }
+  constructor(config?: SeleniumServerProviderConfig) {
+    super();
+    this.cacheFileName = this.setVar('cacheFileName', this.cacheFileName, config);
+    this.configFileName = this.setVar('configFileName', this.configFileName, config);
+    this.ignoreSSL = this.setVar('ignoreSSL', this.ignoreSSL, config);
+    this.osArch = this.setVar('osArch', this.osArch, config);
+    this.osType = this.setVar('osType', this.osType, config);
+    this.outDir = this.setVar('outDir', this.outDir, config);
+    this.port = this.setVar('port', this.port, config);
+    this.proxy = this.setVar('proxy', this.proxy, config);
+    this.requestUrl = this.setVar('requestUrl', this.requestUrl, config);
+    this.runAsNode = this.setVar('runAsNode', this.runAsNode, config);
+    this.runAsDetach = this.setVar('runAsDetach', this.runAsDetach, config);
+    if (this.runAsDetach) {
+      this.runAsNode = true;
     }
   }
 

